core/frontend/src/ViewContext.ts

```diff
@@ -250,7 +250,12 @@ export class DecorateContext extends RenderContext {
     if (this._curCacheableDecorator)
       this._appendToCache({ type: "html", htmlElement: decoration });
 
-    this.screenViewport.decorationDiv.appendChild(decoration);
+    // an element decoration being added might already be on the decorationDiv, just marked for removal
+    if (decoration[ELEMENT_MARKED_FOR_REMOVAL]) {
+      decoration[ELEMENT_MARKED_FOR_REMOVAL] = false;
+    } else if (decoration.parentElement !== this.screenViewport.decorationDiv) {

```

Comment:
> A more serious issue is that any existing decorator that currently does viewport.decorationDiv.appendChild(myElement) is going to start producing exceptions because myElement may have already been appended as a child of decorationDiv on a previous call to decorate. Should probably have tests for this.
@pmconne in addition to moving the tagging here, I added a check for if the element is already attached.
You're saying that there are probably existing (external) decorators appending to the decorationDiv naively, without using `DecorateContext.addHtmlDecoration`? I grepped the code base and didn't see any internal appending of elements to decorationDiv that did not go through `DecorateContext.addHtmlDecoration`. But yes based on the new removal ordering this change will cause errors on people manually appending their own elements when decorating, I'm not sure what to do, the change could cause errors, should we use a try/catch to halt append errors?

Reply:
actually in firefox dev re-appending a child to a container element that already has it as a child doesn't seem to cause any errors or different results

Reply:
> You're saying that there are probably existing (external) decorators appending to the decorationDiv naively, without using `DecorateContext.addHtmlDecoration`?

No I guess I'd be surprised, but of course making `decorationDiv` public (and `@public`) means we can't be certain. Short of monkey-patching all of its methods for adding children to it, I don't think we can do anything about it, other than deprecate `decorationDiv` and advise people to use `addHtmlDecoration`.

Reply:
yeah I don't think we should wrap `decorationDiv` in a proxy or anything, but if we intend to only allow stablely support `addHtmlDecoration`, we could deprecate. Would that turn this into a minor bump change? I'm not sure the semantic versioning rules are for deprecations.

Reply:
Deprecation of a public API is a minor version change in semver. Every commit you make to master targets the next minor version.